### DISCUSSION TITLE
Which SDK should we support next?

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#3489] [Polls]
With the imminent release of the Dagger Go SDK, you can now use Dagger to program your CI/CD pipelines in two different languages: [Go](https://golang.org) and [CUE](https://cuelang.org).

But why stop there? We plan on adding support for many other languages. Our goal is that, no matter what language you want to use to develop your CI/CD pipelines, Dagger has a SDK for it.

Of course, releasing a new SDK is work, so we need to prioritize. To help us prioritize, please 1) vote for which SDK we should support next, and 2) share details on your use case if you can!

If you want to be notified when your SDK is available, you can also [fill this form and share more details about your use case](https://blocklayer.typeform.com/to/a6m5gKSS?typeform-source=github).

Thank you all!

**Comment by maxdeviant** [comment]
I'm looking to use Dagger in our work monorepo. It's all TypeScript using [Rush](https://rushjs.io/) with [pnpm](https://pnpm.io/), so a Dagger Node SDK would be the perfect fit.

We'd be using Dagger for CI and perhaps even bring it into our deployment pipelines as well 🙂

**Comment by tulis** [comment]
A dotnet SDK support would be nice 🙏

**Reply by cdhunt** [reply]
Dotnet could be built upon for Powershell as well.

**Reply by cdhunt** [reply]
https://github.com/cdhunt/dagger/tree/sdk/powershell/sdk/powershell

**Comment by kstkrv** [comment]
I guess Python, Powershell and Bash are the most popular languages among people used to write CI/CD pipelines

**Comment by Alvise88** [comment]
A PureScript support should be great!!!

**Comment by luolong** [comment]
I'd go for TypeScript language next.
Targeting both Node and [Deno](https://deno.land) runtimes.

**Comment by MrChrisRodriguez** [comment]
What's your priority?
- **empower** people that already write pipelines with better tooling?
- **enable** people that don't yet write pipelines but could / should?

If it's the former, then what @kstkrv is right: go to what's already popular. If it's the latter, then JS/TS is the obvious choice. Really comes down to who you're targeting and to what end.

**Reply by jpadams** [reply]
Good question. Definitely both! So we're working on all of them :)

Python SDK is out, as is Golang SDK: https://dagger.io/blog
Node TS/JS SDK is coming very soon as well.
We'll have a good story for Bash to share soon, with more to come.
@cdhunt has been working on a Powershell SDK, which is looking really promising. Hope to see a demo on our next Community call on Thursday, Nov 17 at 10am PT 🎉

**Reply by MrChrisRodriguez** [reply]
Awesome! Do you have any semblance of a timeframe for Node TS/JS (understanding that deadlines slip, etc.)?

**Reply by luolong** [reply]
@jpadams I really hope that the TypeScript backend is capable working with both, NodeJS and Deno backends.

**Reply by shykes** [reply]
> Awesome! Do you have any semblance of a timeframe for Node TS/JS (understanding that deadlines slip, etc.)?

It’s a matter of days 😇

**Reply by shykes** [reply]
> @jpadams I really hope that the TypeScript backend is capable working with both, NodeJS and Deno backends.

We’re focusing on NodeJS first, but would love to add support for Deno. In your experience is that a lot of work? Does it require maintaining two different libraries?

**Reply by luolong** [reply]
I haven't really done any cross framework library work so far and I've only written few small utilities and apps in Deno so far, so I can't speak too much about the fundamental differences between the two, but my impression has been that generally, Deno is less encumbered with the kind of cruft that has been accumulated with Node. 

My impression is that writing code in Deno tends to lean towards writing simply vanilla(ish) TypeScript, while there are definite Node'isms when developing things for Node.

Of course there are Deno and Node specific APIs that are available on one platform but not the other, but barring those, a Deno library might be quite easy to import into Node app, wile importing Node library into Deno app requires some extra effort.

Take it with a grain of salt ... I am only an occasional dabbler in the dark arts of JavaScript/TypeScript development.

**Comment by d3rp3tt3** [comment]
We need a new poll or to update this one since the team has already publicly delivered on several and launched the GraphQL API (this is the API that drives the language-specific SDKs). 

**Docs**
- [Node.js / Typescript / Javascript](https://docs.dagger.io/sdk/nodejs)
- [Python](https://docs.dagger.io/sdk/python)
- [CLI](https://docs.dagger.io/cli)
- [GraphQL API](https://docs.dagger.io/api)

I vote for Rust next. (And I did vote for it.) PowerShell would be compelling as well. The CLI implementation can be used to create a bash script, but having a full Powershell module could unlock a bunch of compelling workflows.

**Comment by sheldonhull** [comment]
Didn't see dotnet  (C#) added there. Thought might be worth mentioning c# despite my love for PowerShell. 😀

**Comment by wingyplus** [comment]
I vote for haskell and another beam languages (erlang, gleam) and zig

**Comment by humrochagf** [comment]
Dotnet core / C# would be nice, it's cross platform now and has a great entry point for corporate.
For example on my personal projects I go always with Python but at work the stack and most of the teams are C# first, and having it as an option helps a lot in the adoption.

**Reply by sheldonhull** [reply]
Are you writing your automation with Nuke or equivalent automation framework?

**Reply by wingyplus** [reply]
I see that dotnet sdk already import to repository as a experimental sdk. But I'm not sure about the state of the sdk. @nabsul may answer better than me. :)

Ref: https://github.com/dagger/dagger/tree/main/sdk/dotnet

**Reply by humrochagf** [reply]
> Are you writing your automation with Nuke or equivalent automation framework?

Depends, some are full gitlab-ci + docker, other pipelines use custom C# automation solution, others teamcity 🙂

**Reply by humrochagf** [reply]
> I see that dotnet sdk already import to repository as a experimental sdk. But I'm not sure about the state of the sdk. @nabsul may answer better than me. :)
> 
> Ref: https://github.com/dagger/dagger/tree/main/sdk/dotnet

Nice, I'll give it a try 🙂

**Reply by nabsul** [reply]
Hi, the .NET SDK is not complete. It has a couple of proof-of-concept examples/ideas, and the beginnings of a GraphQL client code generator. It is not yet usable. I'm not sure when I'll have time to continue work on this, but the project is open to contribution as well :-)

**Reply by edumserrano** [reply]
I also think having .NET SDK support would be great!

**Comment by zeroware** [comment]
Hello, I started working on a PHP SDK. I will try to publish something if there's interest.

**Reply by zeroware** [reply]
Published the first version here : https://github.com/the-diamond-dogs/dagger-php-sdk

**Comment by rechc** [comment]
I come from Jenkins pipelines which written in groovy. I also use maven and write some extended maven goals written in groovy. So I prefer groovy or Java as next supported SDK.

**Reply by wingyplus** [reply]
You can check it out at https://github.com/dagger/dagger/tree/main/sdk/java

**Comment by beingminimal** [comment]
@shykes i think we should work on jsr.io support to make dagger js sdk available for any js runtime like deno and others.
Please have a look. Thanks

**Reply by RyanGuild** [reply]
I would help with this

**Comment by hcnp** [comment]
As a lot of projects are programmed in Java and C# nowadays - it would be cool to use the same language for Dagger.

**Reply by jedevc** [reply]
@hcnp we actually have SDKs for those right now that we're actively working on in collaboration with the community :tada:

You can track the ongoing Java work here: https://github.com/dagger/dagger/issues/9486

For dotnet, I'm not sure we have a tracking issue, but there is a channel on our discord for discussion: https://discord.com/channels/707636530424053791/1193736711801409658

**Reply by wingyplus** [reply]
Here it is https://github.com/dagger/dagger/issues/9391
